source/utils/displayString.py

```diff
@@ -1,64 +1,71 @@
-		display string for a given member in _displayStringLabels.
-		"""
-		pass
-
-	@property
-	def displayString(self) -> str:
-		"""
-		@return: The translated UI display string that should be used for this value of the enum.
-		"""
-		try:
-			return self._displayStringLabels[self]
-		except KeyError:
-			log.error(f"No translation mapping for: {self}")
-			return self._displayStringLabels[self.defaultValue]
+# A part of NonVisual Desktop Access (NVDA)
+# This file is covered by the GNU General Public License.
+# See the file COPYING for more details.
+# Copyright (C) 2021 NV Access Limited.
+
+from abc import ABC, ABCMeta, abstractproperty
+from enum import Enum, EnumMeta, IntEnum
+from typing import Dict
+
+from logHandler import log
+
+
+class _DisplayStringEnumMixinMeta(ABCMeta, EnumMeta):
+	"""
+	This helps correct the Method Resolution Order (MRO) when using the `_DisplayStringEnumMixin`.
+	When creating an Enum with a Mixin, Python suggest an ordering of
+	`class EnumWithMixin(Mixin, type, EnumClass):`.
+	This creates a metaclass conflict as both _DisplayStringEnumMixin and Enum both have metaclasses,
+	ABCMeta and EnumMeta. This requires a new MetaClass which subclasses both of these. This follows the
+	same ordering of the EnumWithMixin usage.
+	See `_DisplayStringEnumMixin`.
+	"""
+	pass
+
+
+class _DisplayStringEnumMixin(ABC):
+	"""
+	This mixin can be used with a class which subclasses Enum to provided translated display strings for
+	members of the enum. The abstract properties must be overridden.
+	To be used with `_DisplayStringEnumMixinMeta`.
+	Usage for python 3.7 is as follows:
+	```
+	class ExampleEnum(_DisplayStringEnumMixin, str, Enum, metaclass=_DisplayStringEnumMixinMeta):
+		pass
+
+	class ExampleIntEnum(_DisplayStringEnumMixin, IntEnum, metaclass=_DisplayStringEnumMixinMeta):
+		pass
+	```
+	"""
+	@abstractproperty
+	def _displayStringLabels(self) -> Dict[Enum, str]:
+		"""
+		Specify a dictionary which takes members of the Enum and returns the translated display string.
+		"""
+		pass
+
+	@property
+	def displayString(self) -> str:
+		"""
+		@return: The translated UI display string that should be used for this value of the enum.
+		"""
+		try:
+			return self._displayStringLabels[self]
+		except KeyError as e:
+			log.error(f"No translation mapping for: {self}")
+			raise e
+
+
+class DisplayStringEnum(_DisplayStringEnumMixin, Enum, metaclass=_DisplayStringEnumMixinMeta):
+	"""An Enum class that adds a displayString property defined by _displayStringLabels"""
+	pass
+
+
+class DisplayStringStrEnum(_DisplayStringEnumMixin, str, Enum, metaclass=_DisplayStringEnumMixinMeta):
+	"""A str Enum class that adds a displayString property defined by _displayStringLabels"""
+	pass
+
+
+class DisplayStringIntEnum(_DisplayStringEnumMixin, IntEnum, metaclass=_DisplayStringEnumMixinMeta):
+	"""An IntEnum class that adds a displayString property defined by _displayStringLabels"""
+	pass

```

Comment:
I've since changed the line endings here to CRLF, I didn't see a need to edit earlier commits to do this as this file doesn't contain moved code.

Reply:
When we normalize line endings, I assume we'll commit as unix line endings IE `LF. Tools like the github editor seem work with unix line endings.

Reply:
I don't think this needs to change by the way. Just an FYI.

Reply:
Hmm, right now most are CRLF I believe.

Reply:
I prefer LF though and would make more developer environments supported I think

Reply:
For what its worth  our [Coding standards](https://github.com/nvaccess/nvda/blob/master/devDocs/codingStandards.md)  currently  states:
>• Most files should contain CRLF line endings, as this is a Windows project and can't be used on Unix-like operating 
systems.

To me this reasoning sounds sane - I don't believe any half decent IDE/editor would have troubles with CRLF  whereas support for LF in Windows only development environment software cannot be taken for granted.

Reply:
Here are some issues I've run into when using VS code and CRLF when working on NVDA

* https://github.com/microsoft/vscode/issues/19254
* https://github.com/microsoft/vscode/issues/123557
* https://github.com/microsoft/vscode/issues/96104

Reply:
Typically a mix of line endings is the real killer. Lots of tools (eg Sublime text) refuse to support it. Both because it is difficult to support, but also on philosophical grounds. Git has several options for line endings, allowing line endings to be converted on checkout / commit. I'm suggesting that we change the line endings to LF when they are committed, which is fairly standard. Much of the git ecosystem is Unix (and thus LF) oriented. When checking out code on windows, developers can have their preference of LF or CRLF.

Despite that NVDA can't be run on unix, it relies on systems that include either unix-like OS or ported unix-like tools.